client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"gopkg.in/errgo.v1"
+)
+
+// Doer is implemented by HTTP client packages
+// to make an HTTP request. It is notably implemented
+// by http.Client and httpbakery.Client.
+//
+// When httprequest uses a Doer value for requests
+// with a non-empty body, it will use DoWithBody if
+// the value implements it (see DoerWithBody).
+// This enables httpbakery.Client to be used correctly.
+type Doer interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// DoerWithBody is implemented by HTTP clients that need
+// to be able to retry HTTP requests with a body.
+// It is notably implemented by httpbakery.Client.
+type DoerWithBody interface {
+	Doer
```

Comment:
I don't think DoerWithBody needs to embed Doer. There's no mechanism that a DoerWithBody that wasn't a Doer could be used.